### PR TITLE
40 Repo implementation in service handlers

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**
Changes:
- Implement CRUD operations for the services in the server handlers
- Added integration tests for `Services` endpoints 
- Added `CreateService` in proto
- Updated proto definition with REST methods
- Changed from `cmux.multiplexer` to `grpc gateway` multiplexer, to fix PATCH request

### CODE REVIEW

File: cmd/translate/main_grpc_integration_test.go [hunk 1]
```diff
@@ -201,3 +204,175 @@ func Test_DownloadTranslationFile_gRPC(t *testing.T) {
...
+	_, err := client.CreateService(ctx, &translatev1.CreateServiceRequest{Service: service})
+	if !assert.NoError(t, err, "Prepare test data") {
+		return
+	}
+
+	service.Name = gofakeit.FirstName()
+
+	_, err = client.UpdateService(ctx, &translatev1.UpdateServiceRequest{
+		Service:    service,
+		UpdateMask: &fieldmaskpb.FieldMask{Paths: []string{"name"}},
```
Comment (VladislavsPerkanuks) [inline cmd/translate/main_grpc_integration_test.go:277]: When adding FieldMask, should it be `service.name`, or just `service` ?

Comment (klavinsjanis) [reply cmd/translate/main_grpc_integration_test.go:277]: for primary resource "name", for nested subresource.name

File: pkg/translate/service.go [hunk 2]
```diff
@@ -9,16 +9,25 @@ import (
...
 )
 
 type (
 	getServiceRequest    translatev1.GetServiceRequest
+	createServiceRequest translatev1.CreateServiceRequest
 	updateServiceRequest translatev1.UpdateServiceRequest
 	deleteServiceRequest translatev1.DeleteServiceRequest
 )
 
+// protoServiceFromService converts model.Service -> translatev1.Service.
```
Comment (klavinsjanis) [inline pkg/translate/service.go:26]: maybe better to use to instead of ->

File: cmd/translate/main.go [hunk 3]
```diff
@@ -125,11 +128,13 @@ func init() {
 	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "translate.yaml", "config file")
 	rootCmd.PersistentFlags().Uint("port", 8080, "port to run service on") //nolint:gomnd
 	rootCmd.PersistentFlags().String("host", "localhost", "host to run service on")
+	rootCmd.PersistentFlags().String("db", "mysql", "db to use with service")
```
Comment (klavinsjanis) [inline cmd/translate/main.go:131]: "database to use ...

File: .vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    {
+      "name": "Launch Package",
+      "type": "go",
+      "request": "launch",
+      "mode": "auto",
+      "program": "${workspaceFolder}/cmd/translate/main.go",
+      "cwd": "${workspaceFolder}",
+      "envFile": "${workspaceFolder}/.env.test",
+      "env": {
+        "OTEL_RESOURCE_ATTRIBUTES": "deployment.environment=debug"
```
Comment (janishorsts) [inline .vscode/launch.json:16]: Do not override existing `deployment.environment`!

File: .vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,20 @@
+{
+  // Use IntelliSense to learn about possible attributes.
+  // Hover to view descriptions of existing attributes.
+  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+  "version": "0.2.0",
+  "configurations": [
+    {
+      "name": "Launch Package",
```
Comment (janishorsts) [inline .vscode/launch.json:8]: ?

```suggestion
      "name": "Debug translate service",
```

Comment (VladislavsPerkanuks) [reply .vscode/launch.json:8]: Yeah, will, this was the default for generating launch.json

File: cmd/translate/main.go [hunk 6]
```diff
@@ -57,10 +71,27 @@ var rootCmd = &cobra.Command{
 		defer grpcServer.GracefulStop()
 
 		mux := runtime.NewServeMux()
-		translatev1.RegisterTranslateServiceServer(grpcServer, &translate.TranslateServiceServer{})
+
+		var (
+			repository repo.Repo
+			errRepo    error
```
Comment (janishorsts) [inline cmd/translate/main.go:77]: Advisory. In this case, you don't have to declare a new error instance. You can reuse `err`.

File: cmd/translate/main_grpc_integration_test.go [hunk 7]
```diff
@@ -201,3 +204,179 @@ func Test_DownloadTranslationFile_gRPC(t *testing.T) {
...
+
+	_, err := client.CreateService(context.Background(), &translatev1.CreateServiceRequest{Service: service})
+	if !assert.NoError(t, err, "Prepare test data") {
+		return
+	}
+
+	actual := status.Code(err)
+	expected := codes.OK
+
+	assert.Equal(t, expected, actual, "want codes.%s got codes.%s", expected, actual)
```
Comment (janishorsts) [inline cmd/translate/main_grpc_integration_test.go:230]: Up to you. But to reduce code boilerplate code (tests by nature are very noisy). It makes it easier to read the tests.

```suggestion
	assert.Equal(t, codes.OK, status.Code(err))
```

Comment (VladislavsPerkanuks) [reply cmd/translate/main_grpc_integration_test.go:230]: I would agree with you, but when tests fail with 
```go
assert.Equal(t, codes.OK, status.Code(err)) 
```
the expected and actual status codes are not human readable, while 
```go
assert.Equal(t, expected, actual, "want codes.%s got codes.%s", expected, actual)
```
On fail prints human-readable expected and actual status code, so I will stick with latter

File: pkg/transform/service.go [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package transform
+
+import (
+	"go.expect.digital/translate/pkg/model"
+	translatev1 "go.expect.digital/translate/pkg/pb/translate/v1"
+)
+
+// ProtoServiceFromService converts model.Service to translatev1.Service.
+func ProtoServiceFromService(service *model.Service) *translatev1.Service {
```
Comment (janishorsts) [inline pkg/transform/service.go:9]: Keep the naming simple. E.g.

```suggestion
func ServiceFromPb(service *translatev1.Service) *model.Service
func ServiceToPb(service *model.Service) *translatev1.Service

// or

func ServiceFromProto(service *translatev1.Service) *model.Service
func ServiceToProto(service *model.Service) *translatev1.Service
```

File: pkg/transform/service.go [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+	return &translatev1.Service{Id: service.ID.String(), Name: service.Name}
```
Comment (janishorsts) [inline pkg/transform/service.go:10]: if uuid.UUID is empty (uuid.Nil), `.String()` returns '00000000-0000-0000-0000-000000000000' but we expect empty string value in the proto message. Set only if it's not uuid.Nil.

File: pkg/transform/service.go
Comment (janishorsts) [inline pkg/transform/service.go:1]: Use property testing for the model transformations

```go
quick.Check(func (id uuid.UUID, name string) {
  service := &model.Service{ID: id, Name: name}

  return assert.Equal(t, service, ServiceFromPb(ServiceToPb(service)))
}, nil)
```

File: pkg/translate/service.go [hunk 10]
```diff
@@ -68,12 +76,78 @@ func (t *TranslateServiceServer) ListServices(
...
+	services, err := t.repo.LoadServices(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	resp := &translatev1.ListServicesResponse{
+		Services: make([]*translatev1.Service, 0, len(services)),
+	}
+
+	for _, v := range services {
```
Comment (janishorsts) [inline pkg/translate/service.go:88]: We can use a generic function for the slice transformation. Please check how it is done in the `tw` project - `transform/transform.go`.

File: pkg/translate/transform.go
Comment (VladislavsPerkanuks) [inline pkg/translate/transform.go:1]: Moved from a separate package to translate pkg.
**Reason**: Only translate pkg needs access to transform (pb -> ours), so there is no need to create separate pkg and create exported functions

File: pkg/translate/service_test.go [hunk 11]
```diff
@@ -4,38 +4,49 @@ import (
...
+	}
+}
+
 func Test_ParseGetServiceParams(t *testing.T) {
 	t.Parallel()
 
+	happyReq := randGetServiceReq()
+
+	malformedIDReq := randGetServiceReq()
+	malformedIDReq.Id += "_FAIL" //nolint:goconst
```
Comment (VladislavsPerkanuks) [inline pkg/translate/service_test.go:30]: Switched to that kind of declaration for test inputs

- Table tests are shorter and more concise
- Random data is generated, which could potentially point to some uncovered edge case

File: pkg/translate/transform.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	return id, nil
+}
+
+// ----------------------Service----------------------
+
+// serviceToProto converts model.Service to translatev1.Service.
+func serviceToProto(s *model.Service) *translatev1.Service {
+	return &translatev1.Service{Id: uuidToProto(s.ID), Name: s.Name}
```
Comment (janishorsts) [inline pkg/translate/transform.go:40]: Cover the case when s is nil when transforming models.

```suggestion
	if s == nil {
		return nil
	}
	
	return &translatev1.Service{Id: uuidToProto(s.ID), Name: s.Name}
```

File: pkg/translate/transform.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	service.ID, err = uuidFromProto(s.Id)
+	if err != nil {
+		return nil, fmt.Errorf("transform id: %w", err)
+	}
+
+	return service, nil
+}
+
+// servicesToProto converts []model.Service to []*translatev1.Service.
+func servicesToProto(s []model.Service) []*translatev1.Service {
```
Comment (janishorsts) [inline pkg/translate/transform.go:59]: Use a generic function to transform a slice. E.g.

```suggestion
func servicesToProto(s []model.Service) []*translatev1.Service { return sliceToProto(s) }

func sliceToProto[T any, R any](slice []T, transform func(*T) *R) []R {
	if len(slice) == 0 {
		return nil
	}

	v := make([]R, 0, len(slice))

	for i := range slice {
		v = append(v, *transform(&slice[i]))
	}

	return v
}
```

File: pkg/translate/transform_test.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+	t.Run("UUID to string to UUID", func(t *testing.T) {
+		t.Parallel()
+
+		f := func(expectedID uuid.UUID) bool {
+			stringID := uuidToProto(expectedID)
+
+			restoredID, err := uuidFromProto(stringID)
+			require.NoError(t, err)
+
+			return assert.Equal(t, expectedID, restoredID)
```
Comment (janishorsts) [inline pkg/translate/transform_test.go:25]:
```suggestion

			return assert.NoError(t, err) && assert.Equal(t, expectedID, restoredID)
```

File: pkg/translate/service.go [hunk 15]
```diff
@@ -68,35 +70,96 @@ func (t *TranslateServiceServer) ListServices(
...
+	return &createServiceParams{service: service}, nil
+}
+
+func (t *TranslateServiceServer) CreateService(
+	ctx context.Context,
+	req *translatev1.CreateServiceRequest,
+) (*translatev1.Service, error) {
+	createReq := (*createServiceRequest)(req)
+
+	params, err := createReq.parseParams()
```
Comment (janishorsts) [inline pkg/translate/service.go:110]: Let's simplify this abstraction - it does not add anything. Consider using functional programming paradigms first.

Please apply it to all request parsing.

```suggestion
	params, err := parseCreateServiceRequest(req)
```

**Review by janishorsts — CHANGES_REQUESTED** [review]
Please be careful what error messages are exposed to the gRPC clients. We do not want to expose internal server errors etc.

PS. I added a ticket to instrument traces in integration tests https://github.com/orgs/translate-agent/projects/1/views/1?pane=issue&itemId=25756421

File: pkg/translate/service.go [hunk 16]
```diff
@@ -152,9 +224,12 @@ func (t *TranslateServiceServer) DeleteService(
 		return nil, status.Errorf(codes.InvalidArgument, err.Error())
 	}
 
-	if err := params.validate(); err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	switch err := t.repo.DeleteService(ctx, params.id); {
+	default:
+		return &emptypb.Empty{}, nil
```
[the remark this answers is not in the document]
Comment (VladislavsPerkanuks) [reply pkg/translate/service.go:229]: Cannot do it. It breaks gRPC

image

File: pkg/translate/service.go [hunk 17]
```diff
@@ -25,41 +23,44 @@ type getServiceParams struct {
...
+		return nil, fmt.Errorf("parse id: %w", err)
 	}
 
-	return getServiceParams{id: id}, nil
+	return &getServiceParams{id: id}, nil
 }
 
-func (g *getServiceParams) validate() error {
-	// validate if uuid is in DB
+func validateGetServiceRequestParams(params *getServiceParams) error {
```
Comment (janishorsts) [inline pkg/translate/service.go:35]: We can leave it as it was previously - `func (g *getServiceParams) validate() error`.

The benefit of `parseGetServiceRequestParams` is it removes the need for the declaration of `params` in a consumer function, and keeps the consumer function tidier.

**Review by janishorsts — APPROVED** [review]
Created a new ticket to validate field mask https://github.com/orgs/translate-agent/projects/1/views/1?pane=issue&itemId=26154387
